flang/lib/Lower/Bridge.cpp

```diff
@@ -1398,89 +1398,137 @@ class FirConverter : public Fortran::lower::AbstractConverter {
+    builder->setInsertionPointToStart(inner.getBody());
     return insPt;
   }
 
-  void genPostlude(bool isHeap, const Fortran::lower::SomeExpr &lhs,
-                   const Fortran::lower::SomeExpr &rhs,
-                   mlir::OpBuilder::InsertPoint insPt) {
-    builder->restoreInsertionPoint(insPt);
-    if (/*copiesWereMade*/ false) {
-      // free buffers
-    }
+  static mlir::Value
+  innerLoopBlockArg(Fortran::lower::ExpressionContext &context) {
+    return context.getArrayBlockArgs().back();
+  }
+  static mlir::Value
+  outerLoopResult(Fortran::lower::ExpressionContext &context) {
+    return context.getLoopReturnVals().front();
+  }
+  mlir::Value genShapeOp(const fir::ExtendedValue &exv) {
+    return Fortran::lower::createShape(toLocation(), *this, exv);
   }
 
-  fir::ExtendedValue
-  genInitializerExprValue(const Fortran::lower::SomeExpr &expr) {
-    Fortran::lower::ExpressionContext context;
-    context.setInInitializer();
-    return createSomeExtendedExpression(toLocation(), *this, expr, localSymbols,
-                                        context);
+  fir::ArrayLoadOp genArrayLoad(Fortran::lower::ExpressionContext &context,
+                                const fir::ExtendedValue &exv) {
+    auto loc = toLocation();
+    auto memref = fir::getBase(exv);
+    auto eleTy = fir::dyn_cast_ptrEleTy(memref.getType());
+    assert(eleTy);
+    auto shape = genShapeOp(exv);
+    mlir::Value slice;
+    return builder->create<fir::ArrayLoadOp>(loc, eleTy, memref, shape, slice,
+                                             llvm::None);
+  }
+
+  void mapArrayExpr(Fortran::lower::ExpressionContext &context,
+                    const Fortran::lower::SomeExpr &expr) {
+    assert(context.inPreludePhase());
+    createSomeExtendedExpression(toLocation(), *this, expr, localSymbols,
+                                 context);
   }
 
-  fir::ExtendedValue genExprEleValue(const Fortran::lower::SomeExpr &expr,
-                                     llvm::ArrayRef<mlir::Value> lcvs) {
+  fir::ExtendedValue
+  genElementalArrayExpr(Fortran::lower::ExpressionContext &context,
+                        const Fortran::lower::SomeExpr &expr) {
+    assert(!context.inPreludePhase());
     return createSomeExtendedExpression(toLocation(), *this, expr, localSymbols,
-                                        lcvs);
+                                        context);
   }
 
-  fir::ExtendedValue genExprEleAddr(const Fortran::lower::SomeExpr &expr,
-                                    llvm::ArrayRef<mlir::Value> lcvs) {
-    return createSomeExtendedAddress(toLocation(), *this, expr, localSymbols,
-                                     lcvs);
+  /// Array assignment.
+  /// 1. Evaluate the lhs to determine the rank and how to form the ArrayLoad
+  /// (e.g., if there is a slicing op).
+  /// 2. Scan the rhs, creating the ArrayLoads and evaluate the scalar subparts
+  /// to be added to the map.
```

Comment:
You're right. This is a hole.

In fact, I want to track these and stick them someplace like the SymbolMap ... but they are not symbols, they are just lowered scalar expressions. I think I will extend the SymbolMap to allow "temporary" bindings. That should be a good general solution applicable to other constructs, etc.

Reply:
The addresses of the `evaluate::Expr` under the parse-tree should be stable, so it may be possible to use these to create a binding here, but that may be a bit risky to assume we do not generate and lower a temporary evaluate::Expr at some point...

A solution could also be to keep a single pass and generate code in both the prelude/loop according to the needs.

We could maybe stick under the Expression entry node (that should also be used while entering operands sub-expressions during the visit) something like:
```
fir::ExtendedValue genval(SomeExpr expr) {
  fir::ExtendedValue res;
  if (inArrayContext()) {
   if (isArray(expr))
     context.generateCodeInLoopNest(); // set builder insertion point
     return genVisit(expr); // lgenVisit would just be like the current genval(SomeExpr)
   else // scalar
     context.generateCodeInPrelude();
     // we are not in an array context for this operand, but we do not want to pop the
    // array context that we still need for the other operands.
     context.setNotInArrayContext(); 
     res = genVisit(expr);
  } else {
    // we are not in array context but if the Expression we want to lower is an elemental operation/function and the expr
    // has an array shape, we need to create a new array context here. This cover for instance the case where a scalar 
    // function is taking an array argument (foo(A + B)) and we need an array context for A + B. 
    if (isElementalExpr(expr)) {
      // we probably want to create a temp array here that can be assigned to, and 
      auto tempArray = context.generateArrayTempStorage(expr);
      context.AddTempToCleanUpList(tempArray);
      context.createNewArrayContext(expr, tempArray);
      auto res = genVisit(expr);
      // pop the array context, the temp cannot be deleted until it has been used;
      context.popArrayContext();
    }
  }
 // clean up the temps created by lower level operand if any (this would be the tempArray that may have been
 //    generated by nested call to `gen(SomeExpr)`, but not the one that we may have just generated in the current
 // call (it's not used yet)).
 context.CleanUpLowerLevelTemp();
}
```

Doing so may also help with the fact that expressions may have several elemental array sub-expressions that should not live in the same array context.

Then, in gen(symbolRef) we would also need to be modified to generate the array loads in the prelude insertion point, and the ref in the loop nest at the same time.

This is just an idea based on the observation that these visits are a central nerve of expression lowering where we could maybe centralized the array expression aspect: https://github.com/flang-compiler/f18-llvm-project/blob/c10d6c987e7d6d341a5486589e2d84483fdb3688/flang/lib/Lower/ConvertExpr.cpp#L1611-L1618
But it may make more sense to deploy some sort of common handling in the expressions node that are elemental instead.

Reply:
I'm reluctant to push too much of the smarts into the expression helper. I don't think that's the right place for it after all. I'm going to experiment a bit...